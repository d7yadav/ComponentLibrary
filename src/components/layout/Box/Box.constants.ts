import { 
  BoxDisplay, 
  BoxPosition, 
  BoxOverflow, 
  BoxTextAlign, 
  BoxVerticalAlign,
  BoxFlexDirection,
  BoxFlexWrap,
  BoxJustifyContent,
  BoxAlignItems,
  BoxAlignContent
} from './Box.types';

/**
 * BOX_DISPLAYS component
 * 
 * @returns JSX element
 */
export const BOX_DISPLAYS: Record<BoxDisplay, BoxDisplay> = {
  block: 'block',
  inline: 'inline',
  'inline-block': 'inline-block',
  flex: 'flex',
  'inline-flex': 'inline-flex',
  grid: 'grid',
  'inline-grid': 'inline-grid',
  none: 'none',
  contents: 'contents',
} as const;

/**
 * BOX_POSITIONS component
 * 
 * @returns JSX element
 */
export const BOX_POSITIONS: Record<BoxPosition, BoxPosition> = {
  static: 'static',
  relative: 'relative',
  absolute: 'absolute',
  fixed: 'fixed',
  sticky: 'sticky',
} as const;

/**
 * BOX_OVERFLOWS component
 * 
 * @returns JSX element
 */
export const BOX_OVERFLOWS: Record<BoxOverflow, BoxOverflow> = {
  visible: 'visible',
  hidden: 'hidden',
  scroll: 'scroll',
  auto: 'auto',
} as const;

/**
 * BOX_TEXT_ALIGNS component
 * 
 * @returns JSX element
 */
export const BOX_TEXT_ALIGNS: Record<BoxTextAlign, BoxTextAlign> = {
  left: 'left',
  center: 'center',
  right: 'right',
  justify: 'justify',
  start: 'start',
  end: 'end',
} as const;

/**
 * BOX_VERTICAL_ALIGNS component
 * 
 * @returns JSX element
 */
export const BOX_VERTICAL_ALIGNS: Record<BoxVerticalAlign, BoxVerticalAlign> = {
  baseline: 'baseline',
  top: 'top',
  middle: 'middle',
  bottom: 'bottom',
  'text-top': 'text-top',
  'text-bottom': 'text-bottom',
} as const;

/**
 * BOX_FLEX_DIRECTIONS component
 * 
 * @returns JSX element
 */
export const BOX_FLEX_DIRECTIONS: Record<BoxFlexDirection, BoxFlexDirection> = {
  row: 'row',
  'row-reverse': 'row-reverse',
  column: 'column',
  'column-reverse': 'column-reverse',
} as const;

/**
 * BOX_FLEX_WRAPS component
 * 
 * @returns JSX element
 */
export const BOX_FLEX_WRAPS: Record<BoxFlexWrap, BoxFlexWrap> = {
  nowrap: 'nowrap',
  wrap: 'wrap',
  'wrap-reverse': 'wrap-reverse',
} as const;

/**
 * BOX_JUSTIFY_CONTENTS component
 * 
 * @returns JSX element
 */
export const BOX_JUSTIFY_CONTENTS: Record<BoxJustifyContent, BoxJustifyContent> = {
  'flex-start': 'flex-start',
  'flex-end': 'flex-end',
  center: 'center',
  'space-between': 'space-between',
  'space-around': 'space-around',
  'space-evenly': 'space-evenly',
} as const;

/**
 * BOX_ALIGN_ITEMS component
 * 
 * @returns JSX element
 */
export const BOX_ALIGN_ITEMS: Record<BoxAlignItems, BoxAlignItems> = {
  'flex-start': 'flex-start',
  'flex-end': 'flex-end',
  center: 'center',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;

/**
 * BOX_ALIGN_CONTENTS component
 * 
 * @returns JSX element
 */
export const BOX_ALIGN_CONTENTS: Record<BoxAlignContent, BoxAlignContent> = {
  'flex-start': 'flex-start',
  'flex-end': 'flex-end',
  center: 'center',
  'space-between': 'space-between',
  'space-around': 'space-around',
  stretch: 'stretch',
} as const;

/**
 * BOX_BREAKPOINTS component
 * 
 * @returns JSX element
 */
export const BOX_BREAKPOINTS = {
  xs: 0,
  sm: 600,
  md: 900,
  lg: 1200,
  xl: 1536,
} as const;

/**
 * BOX_SPACING_VALUES component
 * 
 * @returns JSX element
 */
export const BOX_SPACING_VALUES = {
  0: 0,
  0.5: 0.5,
  1: 1,
  1.5: 1.5,
  2: 2,
  2.5: 2.5,
  3: 3,
  3.5: 3.5,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  10: 10,
  11: 11,
  12: 12,
  14: 14,
  16: 16,
  20: 20,
  24: 24,
  28: 28,
  32: 32,
  36: 36,
  40: 40,
  44: 44,
  48: 48,
  52: 52,
  56: 56,
  60: 60,
  64: 64,
  72: 72,
  80: 80,
  96: 96,
} as const;

/**
 * BOX_BORDER_RADIUS component
 * 
 * @returns JSX element
 */
export const BOX_BORDER_RADIUS = {
  none: 0,
  sm: 4,
  md: 8,
  lg: 12,
  xl: 16,
  '2xl': 20,
  '3xl': 24,
  full: '50%',
} as const;

/**
 * BOX_SHADOW_VALUES component
 * 
 * @returns JSX element
 */
export const BOX_SHADOW_VALUES = {
  none: 0,
  sm: 1,
  md: 3,
  lg: 6,
  xl: 12,
  '2xl': 24,
} as const;

/**
 * ACCESSIBILITY_CONSTANTS component
 * 
 * @returns JSX element
 */
export const ACCESSIBILITY_CONSTANTS = {
  regionRole: 'region',
  bannerRole: 'banner',
  mainRole: 'main',
  contentinfoRole: 'contentinfo',
  complementaryRole: 'complementary',
  navigationRole: 'navigation',
  articleRole: 'article',
  sectionRole: 'section',
  asideRole: 'aside',
  figureRole: 'figure',
  groupRole: 'group',
  presentationRole: 'presentation',
  noneRole: 'none',
} as const;

/**
 * COMMON_BOX_PATTERNS component
 * 
 * @returns JSX element
 */
export const COMMON_BOX_PATTERNS = {
  // Layout patterns
  container: {
    maxWidth: 'lg',
    mx: 'auto',
    px: 2,
  },
  
  section: {
    py: 8,
    px: 2,
  },
  
  card: {
    p: 3,
    border: 1,
    borderRadius: 2,
    bgcolor: 'background.paper',
    boxShadow: 2,
  },
  
  hero: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: '50vh',
    textAlign: 'center',
    p: 6,
  },
  
  sidebar: {
    width: 250,
    height: '100vh',
    borderRight: 1,
    borderColor: 'divider',
    p: 2,
  },
  
  // Flexbox patterns
  flexCenter: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  
  flexBetween: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  
  flexColumn: {
    display: 'flex',
    flexDirection: 'column',
  },
  
  flexWrap: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  
  // Grid patterns
  gridCenter: {
    display: 'grid',
    placeItems: 'center',
  },
  
  gridTwoColumns: {
    display: 'grid',
    gridTemplateColumns: 'repeat(2, 1fr)',
    gap: 2,
  },
  
  gridThreeColumns: {
    display: 'grid',
    gridTemplateColumns: 'repeat(3, 1fr)',
    gap: 2,
  },
  
  gridResponsive: {
    display: 'grid',
    gridTemplateColumns: {
      xs: '1fr',
      sm: 'repeat(2, 1fr)',
      md: 'repeat(3, 1fr)',
      lg: 'repeat(4, 1fr)',
    },
    gap: 2,
  },
  
  // Position patterns
  absoluteCenter: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
  },
  
  fixed: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    zIndex: 1000,
  },
  
  sticky: {
    position: 'sticky',
    top: 0,
    zIndex: 10,
  },
  
  // Utility patterns
  fullscreen: {
    width: '100vw',
    height: '100vh',
  },
  
  aspectRatio: {
    position: 'relative',
    overflow: 'hidden',
    '&::before': {
      content: '""',
      display: 'block',
      paddingTop: '56.25%', // 16:9 aspect ratio
    },
  },
  
  overlay: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    bgcolor: 'rgba(0, 0, 0, 0.5)',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
} as const;

/**
 * RESPONSIVE_PATTERNS component
 * 
 * @returns JSX element
 */
export const RESPONSIVE_PATTERNS = {
  // Hide/show patterns
  hideOnMobile: {
    display: { xs: 'none', sm: 'block' },
  },
  
  hideOnDesktop: {
    display: { xs: 'block', sm: 'none' },
  },
  
  showOnMobile: {
    display: { xs: 'block', sm: 'none' },
  },
  
  showOnDesktop: {
    display: { xs: 'none', sm: 'block' },
  },
  
  // Responsive sizing
  responsiveWidth: {
    width: { xs: '100%', sm: 'auto' },
  },
  
  responsivePadding: {
    p: { xs: 2, sm: 3, md: 4, lg: 6 },
  },
  
  responsiveMargin: {
    m: { xs: 1, sm: 2, md: 3, lg: 4 },
  },
  
  // Responsive typography
  responsiveText: {
    fontSize: { xs: '1rem', sm: '1.125rem', md: '1.25rem', lg: '1.5rem' },
  },
} as const;

/**
 * DEFAULT_BOX_PROPS component
 * 
 * @returns JSX element
 */
export const DEFAULT_BOX_PROPS = {
  component: 'div',
  centered: false,
  rounded: false,
  elevated: false,
  fullWidth: false,
  fullHeight: false,
  clickable: false,
} as const;