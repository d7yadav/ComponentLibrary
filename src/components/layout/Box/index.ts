export { Box } from './Box';
export type { 
  BoxProps, 
  BoxDisplay, 
  BoxPosition,
  BoxOverflow,
  BoxTextAlign,
  BoxVerticalAlign,
  BoxFlexDirection,
  BoxFlexWrap,
  BoxJustifyContent,
  BoxAlignItems,
  BoxAlignContent,
  ResponsiveValue,
  BoxStyleProps,
  BoxBreakpoints
} from './Box.types';
export {
  BOX_DISPLAYS,
  BOX_POSITIONS,
  BOX_OVERFLOWS,
  BOX_TEXT_ALIGNS,
  BOX_VERTICAL_ALIGNS,
  BOX_FLEX_DIRECTIONS,
  BOX_FLEX_WRAPS,
  BOX_JUSTIFY_CONTENTS,
  BOX_ALIGN_ITEMS,
  BOX_ALIGN_CONTENTS,
  BOX_BREAKPOINTS,
  BOX_SPACING_VALUES,
  BOX_BORDER_RADIUS,
  BOX_SHADOW_VALUES,
  ACCESSIBILITY_CONSTANTS,
  COMMON_BOX_PATTERNS,
  RESPONSIVE_PATTERNS,
  DEFAULT_BOX_PROPS,
} from './Box.constants';
export {
  FlexBox,
  FlexCenterBox,
  FlexBetweenBox,
  FlexColumnBox,
  GridBox,
  GridCenterBox,
  CardBox,
  HeroBox,
  SectionBox,
  ContainerBox,
  SidebarBox,
  OverlayBox,
  AspectRatioBox,
  StickyBox,
  ScrollableBox,
  ClickableBox,
} from './Box.styles';