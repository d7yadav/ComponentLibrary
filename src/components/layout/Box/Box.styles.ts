import { Box as MuiBox } from '@mui/material';
import { styled } from '@mui/material/styles';
import React from 'react';

import { BOX_BORDER_RADIUS, BOX_SHADOW_VALUES } from './Box.constants';
export const StyledBox: React.ComponentType<any> = styled(MuiBox, {
  shouldForwardProp: (prop) => !['customCentered', 'customRounded', 'customElevated', 'customFullWidth', 'customFullHeight', 'customClickable'].includes(prop as string),
})<any>((props: any) => {
  const { 
    theme, 
    customCentered,
    customRounded,
    customElevated,
    customFullWidth,
    customFullHeight,
    customClickable
  } = props;
  
  return {
    // Centered content
    ...(customCentered && {
      'display': 'flex',
      'alignItems': 'center',
      'justifyContent': 'center',
    }),
    
    // Border radius
    ...(customRounded && {
      'borderRadius': typeof customRounded === 'number' 
        ? customRounded 
        : BOX_BORDER_RADIUS.md,
    }),
    
    // Elevation (box shadow)
    ...(customElevated && {
      'boxShadow': typeof customElevated === 'number' 
        ? theme.shadows[Math.min(customElevated, 24)]
        : theme.shadows[BOX_SHADOW_VALUES.md],
    }),
    
    // Full width
    ...(customFullWidth && {
      'width': '100%',
    }),
    
    // Full height
    ...(customFullHeight && {
      'height': '100%',
    }),
    
    // Clickable styles
    ...(customClickable && {
      'cursor': 'pointer',
      'transition': theme.transitions.create(
        ['background-color', 'box-shadow', 'transform'],
        {
          duration: theme.transitions.duration.short,
          easing: theme.transitions.easing.easeInOut,
        }
      ),
      
      '&:hover': {
        'backgroundColor': theme.palette.action.hover,
        'transform': 'translateY(-1px)',
        'boxShadow': theme.shadows[4],
      },
      
      '&:active': {
        'transform': 'translateY(0)',
        'boxShadow': theme.shadows[2],
      },
      
      '&:focus-visible': {
        'outline': `2px solid ${theme.palette.primary.main}`,
        'outlineOffset': '2px',
      },
    }),
  };
});

// Specialized Box components for common patterns
export const FlexBox: React.ComponentType<any> = styled(StyledBox)<any>(() => ({
  'display': 'flex',
}));

export const FlexCenterBox: React.ComponentType<any> = styled(FlexBox)<any>(() => ({
  'alignItems': 'center',
  'justifyContent': 'center',
}));

export const FlexBetweenBox: React.ComponentType<any> = styled(FlexBox)<any>(() => ({
  'alignItems': 'center',
  'justifyContent': 'space-between',
}));

export const FlexColumnBox: React.ComponentType<any> = styled(FlexBox)<any>(() => ({
  'flexDirection': 'column',
}));

export const GridBox: React.ComponentType<any> = styled(StyledBox)<any>(() => ({
  'display': 'grid',
}));

export const GridCenterBox: React.ComponentType<any> = styled(GridBox)<any>(() => ({
  'placeItems': 'center',
}));

export const CardBox: React.ComponentType<any> = styled(StyledBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'padding': theme.spacing(3),
    'borderRadius': BOX_BORDER_RADIUS.md,
    'boxShadow': theme.shadows[2],
    'backgroundColor': theme.palette.background.paper,
    'border': `1px solid ${theme.palette.divider}`,
  };
});

export const HeroBox: React.ComponentType<any> = styled(FlexCenterBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'minHeight': '50vh',
    'padding': theme.spacing(6),
    'textAlign': 'center',
    'flexDirection': 'column',
    
    [theme.breakpoints.down('md')]: {
      'minHeight': '40vh',
      'padding': theme.spacing(4),
    },
  };
});

export const SectionBox: React.ComponentType<any> = styled(StyledBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'paddingTop': theme.spacing(8),
    'paddingBottom': theme.spacing(8),
    'paddingLeft': theme.spacing(2),
    'paddingRight': theme.spacing(2),
    
    [theme.breakpoints.up('md')]: {
      'paddingTop': theme.spacing(12),
      'paddingBottom': theme.spacing(12),
      'paddingLeft': theme.spacing(4),
      'paddingRight': theme.spacing(4),
    },
  };
});

export const ContainerBox: React.ComponentType<any> = styled(StyledBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'maxWidth': theme.breakpoints.values.lg,
    'marginLeft': 'auto',
    'marginRight': 'auto',
    'paddingLeft': theme.spacing(2),
    'paddingRight': theme.spacing(2),
    
    [theme.breakpoints.up('sm')]: {
      'paddingLeft': theme.spacing(3),
      'paddingRight': theme.spacing(3),
    },
  };
});

export const SidebarBox: React.ComponentType<any> = styled(StyledBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'width': 250,
    'height': '100vh',
    'borderRight': `1px solid ${theme.palette.divider}`,
    'padding': theme.spacing(2),
    'backgroundColor': theme.palette.background.paper,
    'overflowY': 'auto',
    
    [theme.breakpoints.down('md')]: {
      'width': '100%',
      'height': 'auto',
      'borderRight': 'none',
      'borderBottom': `1px solid ${theme.palette.divider}`,
    },
  };
});

export const OverlayBox: React.ComponentType<any> = styled(StyledBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'position': 'absolute',
    'top': 0,
    'left': 0,
    'right': 0,
    'bottom': 0,
    'backgroundColor': 'rgba(0, 0, 0, 0.5)',
    'display': 'flex',
    'alignItems': 'center',
    'justifyContent': 'center',
    'zIndex': theme.zIndex.modal,
  };
});

export const AspectRatioBox: React.ComponentType<any> = styled(StyledBox, {
  shouldForwardProp: (prop) => !['aspectRatio'].includes(prop as string),
})<{ aspectRatio?: number }>((props: any) => {
  const { aspectRatio = 16 / 9 } = props;
  const paddingTop = `${(1 / aspectRatio) * 100}%`;
  
  return {
    'position': 'relative',
    'overflow': 'hidden',
    
    '&::before': {
      'content': '""',
      'display': 'block',
      'paddingTop': paddingTop,
    },
    
    '& > *': {
      'position': 'absolute',
      'top': 0,
      'left': 0,
      'width': '100%',
      'height': '100%',
      'objectFit': 'cover',
    },
  };
});

export const StickyBox: React.ComponentType<any> = styled(StyledBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'position': 'sticky',
    'top': 0,
    'zIndex': theme.zIndex.appBar,
    'backgroundColor': theme.palette.background.paper,
    'borderBottom': `1px solid ${theme.palette.divider}`,
  };
});

export const ScrollableBox: React.ComponentType<any> = styled(StyledBox)<any>(() => ({
  'overflow': 'auto',
  
  // Custom scrollbar styling
  '&::-webkit-scrollbar': {
    'width': 8,
    'height': 8,
  },
  
  '&::-webkit-scrollbar-track': {
    'backgroundColor': 'transparent',
  },
  
  '&::-webkit-scrollbar-thumb': {
    'backgroundColor': 'rgba(0, 0, 0, 0.2)',
    'borderRadius': 4,
    
    '&:hover': {
      'backgroundColor': 'rgba(0, 0, 0, 0.3)',
    },
  },
}));

export const ClickableBox: React.ComponentType<any> = styled(StyledBox)<any>((props: any) => {
  const { theme } = props;
  return {
    'cursor': 'pointer',
    'transition': theme.transitions.create(
      ['background-color', 'box-shadow', 'transform'],
      {
        duration: theme.transitions.duration.short,
        easing: theme.transitions.easing.easeInOut,
      }
    ),
    
    '&:hover': {
      'backgroundColor': theme.palette.action.hover,
      'transform': 'translateY(-2px)',
      'boxShadow': theme.shadows[4],
    },
    
    '&:active': {
      'transform': 'translateY(0)',
      'boxShadow': theme.shadows[2],
    },
    
    '&:focus-visible': {
      'outline': `2px solid ${theme.palette.primary.main}`,
      'outlineOffset': '2px',
    },
  };
});