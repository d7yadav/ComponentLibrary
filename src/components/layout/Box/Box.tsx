import { forwardRef, memo } from 'react';
import { BoxProps } from './Box.types';
import { StyledBox } from './Box.styles';
import {
  DEFAULT_BOX_PROPS,
} from './Box.constants';

/**
 * Enhanced Box component with general-purpose styling props and responsive support
 * 
 * Features:
 * - Comprehensive styling system with spacing, sizing, positioning
 * - Responsive prop support for all style properties
 * - Flexbox and CSS Grid utilities built-in
 * - Color and typography prop shortcuts
 * - Border and shadow utility props
 * - Custom styling shortcuts (centered, rounded, elevated, etc.)
 * - Accessibility support with proper ARIA attributes
 * - TypeScript support with strict typing
 * - Dark theme support via enhanced theme system
 * - Performance optimized with shouldForwardProp
 */
export const Box = forwardRef<HTMLDivElement, BoxProps>((
  {
    children,
    component = DEFAULT_BOX_PROPS.component,
    // Custom shortcuts
    centered = DEFAULT_BOX_PROPS.centered,
    rounded = DEFAULT_BOX_PROPS.rounded,
    elevated = DEFAULT_BOX_PROPS.elevated,
    fullWidth = DEFAULT_BOX_PROPS.fullWidth,
    fullHeight = DEFAULT_BOX_PROPS.fullHeight,
    clickable = DEFAULT_BOX_PROPS.clickable,
    // Standard props
    className,
    sx,
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    role,
    tabIndex,
    ...other
  },
  ref
) => {
  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    role,
    ...(tabIndex !== undefined && { tabIndex }),
    // Add keyboard support for clickable boxes
    ...(clickable && {
      tabIndex: tabIndex !== undefined ? tabIndex : 0,
      role: role || 'button',
      onKeyDown: (event: React.KeyboardEvent) => {
        if (event.key === 'Enter' || event.key === ' ') {
          event.preventDefault();
          (event.target as HTMLElement).click();
        }
      },
    }),
  };

  return (
    <StyledBox data-testid="box"
      ref={ref}
      component={component}
      className={className}
      sx={sx}
      // Custom props for styled component
      customCentered={centered}
      customRounded={rounded}
      customElevated={elevated}
      customFullWidth={fullWidth}
      customFullHeight={fullHeight}
      customClickable={clickable}
      {...accessibilityProps}
      {...other}
    >
      {children}
    </StyledBox>
  );
});

Box.displayName = 'Box';