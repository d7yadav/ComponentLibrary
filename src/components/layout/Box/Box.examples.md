# ðŸ“– Box Examples

## Basic Usage

```typescript
import { Box } from '@/components/layout/Box';

// Basic usage
const Example = () => {
  return (
    <Box>
      Basic Box example
    </Box>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Box
      variant="primary"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Box
    </Box>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Box
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Box disabled>
    No content available
  </Box>
);

// Edge case: Error states
const ErrorState = () => (
  <Box 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Box 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedBox = React.memo((props: BoxProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Box 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Box key={index}>
          Item {index}
        </Box>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Box
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Box
    </Box>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Box
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Box>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Box component.*
