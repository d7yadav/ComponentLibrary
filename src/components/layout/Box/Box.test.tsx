/**
 * @fileoverview Box Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Box } from './Box';
import { BOX_CONSTANTS } from './Box.constants';
import type { BoxProps } from './Box.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Box Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Box>Test content</Box>);
      expect(screen.getByTestId('box-root')).toBeInTheDocument();
    });

    it('renders children correctly', () => {
      const testContent = 'Test Box Content';
      renderWithTheme(<Box>{testContent}</Box>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-box';
      renderWithTheme(
        <Box data-testid={testId}>Test</Box>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Box className={customClass}>Test</Box>
      );
      expect(screen.getByTestId('box-root')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: BoxProps['variant'][] = [
        'default',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Box variant={variant} data-testid="variant-test">
              {variant} variant
            </Box>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: BoxProps['size'][] = [
        'auto',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Box size={size} data-testid="size-test">
              {size} size
            </Box>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });



    it('handles disabled state correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Box disabled onClick={handleClick} data-testid="disabled-test">
          Disabled Box
        </Box>
      );
      
      const element = screen.getByTestId('disabled-test');
      expect(element).toHaveAttribute('aria-disabled', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles click events when clickable', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Box onClick={handleClick} data-testid="clickable-box">
          Clickable Box
        </Box>
      );
      
      const element = screen.getByTestId('clickable-box');
      fireEvent.click(element);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('supports keyboard interactions when focusable', () => {
      renderWithTheme(
        <Box tabIndex={0} data-testid="focusable-box">
          Focusable Box
        </Box>
      );
      
      const element = screen.getByTestId('focusable-box');
      element.focus();
      expect(element).toHaveFocus();
    });
  });



  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Box>Accessible Box</Box>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Box aria-label={ariaLabel}>
          Box for screen readers
        </Box>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<Box>Keyboard Box</Box>);
      
      const element = screen.getByRole('generic');
      element.focus();
      expect(element).toHaveFocus();
          });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Box 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Box
        </Box>
      );
      
      const element = screen.getByTestId('box-root');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiBox: {
            defaultProps: {
              variant: 'default',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Box data-testid="themed-box">
            Themed Box
          </Box>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-box');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Box>Dark Mode Box</Box>
        </ThemeProvider>
      );

      expect(screen.getByTestId('box-root')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Box key={i}>Box {i}</Box>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[data-testid*="box"]')).toHaveLength(100);
    });

  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Box />);
      expect(screen.getByTestId('box-root')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Box>{null}</Box>);
      expect(screen.getByTestId('box-root')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Box 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </Box>
      );
      expect(screen.getByTestId('box-root')).toBeInTheDocument();
    });

  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Box>Default Box</Box>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: BoxProps['variant'][] = [
        'default',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Box variant={variant}>{variant} Box</Box>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
