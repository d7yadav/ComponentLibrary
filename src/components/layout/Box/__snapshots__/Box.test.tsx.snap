// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Box Component > Snapshots > matches snapshot for all variants > variant-default 1`] = `
<div
  class="MuiBox-root css-1xdhyk6"
  data-testid="box"
  variant="default"
>
  default
   Box
</div>
`;

exports[`Box Component > Snapshots > matches snapshot for default props 1`] = `
<div
  class="MuiBox-root css-1xdhyk6"
  data-testid="box"
>
  Default Box
</div>
`;
